axes/decorators.py

```diff
@@ -160,15 +160,11 @@ def query2str(items, max_length=1024):
 
     If there's a field called "password" it will be excluded from the output.
 
-    The length of the output is limited to max_length to avoid a DoS attack.
+    The length of the output is limited to max_length to avoid a DoS attack via excessively large payloads.
     """
 
-    kvs = []
-    for k, v in items:
-        if k != PASSWORD_FORM_FIELD:
-            kvs.append(six.u('%s=%s') % (k, v))
-
-    return '\n'.join(kvs)[:max_length]
+    return '\n'.join(['%s=%s' % (k, v) for k, v in six.iteritems(items)
+                      if k != PASSWORD_FORM_FIELD][:int(max_length/2)])[:max_length]
```

Comment:
That's a minor optimization to limit the number of elements passed to the join(), in case of a GET or POST payload with an excessive number of elements. It's a little shortcut, since the end result is truncated to `max_length` anyways. Actually, this could probably even be `int(max_length/4)`. I'm having trouble properly articulating it right now (brain malfunction) but, given that the output is in the format `k=v\n`, the most elements which could fit into the eventual output is `max_length/4` (4 characters per element).